mernarezk/assignment2/Q1.java

```diff
@@ -0,0 +1,78 @@
+package Assignment2;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.swing.plaf.synth.SynthSpinnerUI;
+
+public class Q1 {
+	/**
+	 * Method that gets all possible paths in a Binarty Tree, then traverses the
+	 * paths one by one and tries to find our target value in the lists ( since
+	 * there are no duplicates) once our target value was found we return a sublist
+	 * of that path from root till target.
+	 * 
+	 * @param root
+	 * @param target
+	 *            val to get it's ancestors
+	 * @return a list of all ancestors for given target
+	 * @throws IllegalStateException
+	 */
+	public static List<Integer> getAncestors(TreeNode root, int target) throws IllegalStateException {

```

Comment:
What would you say is the computational and memory complexity of this method? How would you solve this problem if you stored the reference to a parent in a Node?

Reply:
Ya I figured out it used lots of spaces the runtime would be O(NLogN) since I am getting all paths and space complexity would be (2^N)?? I am not sure .. But I changed the solution to use the parent pointer which will take O(N) for searching the whole tree for the target :) and space would be lg(N) since the worse case of a path would be the height of the tree from root to leaf. correct me if I am wrong!

Reply:
Very nice. :) You can either remove the old solution or add a deprecation warning to its javadoc.